activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb

```diff
@@ -1241,7 +1241,10 @@ def index_name_for_remove(table_name, options = {})
           end
 
           if column_names.present?
-            checks << lambda { |i| index_name(table_name, i.columns) == index_name(table_name, column_names) }
+            checks << lambda do |i|
+              index_name(table_name, i.columns) == index_name(table_name, column_names) ||
+                i.name == index_name(table_name, column_names)
```

Comment:
Thanks for the new spec.  

I'm not sure if both of these specs can be satisfied with a reasonable modification to the remove index code.  The ambiguity that allows `connection.remove_index("testings", "CAST(last_name AS text)")` to lookup the same index will not pass the spec in https://github.com/rails/rails/commit/e24eeef3d42eac4f6fdd9a63bcde2a054e839af9

Does it make sense to modify the code under `add_index` so that "index_payments_on_CAST_created_at_AS_date" is instead called "index_payments_on_created_at_date" from the very beginning?  I believe doing so would not be backwards compatible though?  So this fix could only be applied for version 6.0?